core/interval/int.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package interval
+
+import (
+	"sort"
+)
+
+type IntegerInterval struct {
+	Lower int
+	Upper int
+}
+
+func NewIntegerInterval(a, b int) IntegerInterval {
```

Comment:
Arguably if a>b then the interval should be empty, that would match the mathematical interpretation on (a,b).

Reply:
I have gone back and forth on this a few time. I _think_ this is the best of many non-perfect solutions.

The options I have evaluated are:

1) Having params `(lower, upper int)` and returning an error if lower > upper. However, returning an error here makes things ugly in other places.

2) Having paramsd `(lower, upper int)`, returning no error, and in the case where lower > upper consider this value to be the empty set.
This is not very nice. it's unexpected and error prone.
We also have the case where, mathematically, `[2, 1] = {} = [3, 0]` but `IntegerInterval{2, 1} != IntegerInterval{3, 0}`
Imo the empty set should be represented as an empty `IntegerIntervals`, not an `IntegerInterval` proper. 

3) The implemented solution. This has the problems described above, but the advantage of no errors when constructing and a cleaner use case when the user doesn't know which of their params are greater.

Reply:
This is defo fine now it has been made specific to port ranges